databuilder/typing_utils.py

```diff
@@ -0,0 +1,143 @@
+
+        `dict[str, FileNotFoundError]` matches `dict[Any, OSError]`
+
+    Because `str` matches `Any` and `FileNotFoundError` is a subclass of `OSError`.
+
+    The specification may contain TypeVars. The first match for each variable will bind
+    its value, recording it in the `typevar_context` dict. Subsequent uses of the
+    variable will fail to match if they don't have the same type. By sharing the
+    `typevar_context` dict across different calls to this function you can enforce
+    consistent interpretation of the TypeVar.
+
+    For example, both the below calls will match:
+
+        T = TypeVar("T")
+        type_matches(dict[int, int], dict[T, T], {})
+        type_matches(dict[str, str], dict[T, T], {})
+
+    But these will fail:
+
+        ctx = {}
+        type_matches(dict[int, bool], dict[T, T], ctx)
+        type_matches(dict[str, str], dict[T, T], ctx)
+
+    The first because the value of T is internally inconsistent; the second because
+    although it is internally consistent it is inconsistent with the previous example
+    whose context it shares.
+    """
+    # If `spec` is a type variable
+    if isinstance(spec, typing.TypeVar):
+        # Check the type constraints are met, if any
+        if spec.__constraints__:
+            if not any(
+                type_matches(type_, subspec, typevar_context)
+                for subspec in spec.__constraints__
+            ):
+                return False
+        # If we've already assigned a value for this variable (and it wasn't the Any
+        # type) then check it matches
+        if spec in typevar_context and typevar_context[spec] != typing.Any:
+            if typevar_context[spec] != type_:
+                return False
+        # Otherwise record this value as the value of the TypeVar
+        else:
+            typevar_context[spec] = type_
+        return True
+
+    # `Any` is the easy case: it just matches everything
+    if type_ is typing.Any or spec is typing.Any:
+        return True
+
+    spec_origin = typing.get_origin(spec)
+    spec_args = typing.get_args(spec)
+
+    # If there's no origin type that means `spec` is an ordinary class
+    if spec_origin is None:
+        return type_ is not None and issubclass(type_, spec)
+    elif spec_origin is typing.Union:
+        # For union types we just need to match one of the arguments
+        return any(type_matches(type_, arg, typevar_context) for arg in spec_args)
+    else:
+        # Otherwise we get origin and args for `type_` and check that each element
+        # matches
+        type_origin = typing.get_origin(type_)
+        if not type_matches(type_origin, spec_origin, typevar_context):
+            return False
+        type_args = typing.get_args(type_)
+        for type_arg, spec_arg in zip(type_args, spec_args):
+            if not type_matches(type_arg, spec_arg, typevar_context):
+                return False
+        return True
+
+
+@singledispatch
+def get_typespec(value):
```

Comment:
Yeah, we could just take a fixed list of types here and error on anything else. This gets a bit awkward because one of the types we need to accept is `Code`, but we can't import that here without circularity so we need to register a new `singledispatch` function for it. Which is fine, I was just being a bit lazy. But I think the safety is worth it, isn't it?

Reply:
I've found a better way: we can assert the type doesn't take arguments, as you originally suggested